src/logic/function/basic.lean

```diff
@@ -79,6 +80,20 @@ def injective.decidable_eq [decidable_eq β] (I : injective f) : decidable_eq α
 lemma injective.of_comp {g : γ → α} (I : injective (f ∘ g)) : injective g :=
 λ x y h, I $ show f (g x) = f (g y), from congr_arg f h
 
+lemma injective.dite_injective (p : α → Prop) [decidable_pred p]
+  (f : {a : α // p a} → β) (f' : {a : α // ¬ p a} → β)
+  (hf : function.injective f) (hf' : function.injective f')
+  (im_disj : ∀ {x x' : α} {hx : p x} {hx' : ¬ p x'}, f ⟨x, hx⟩ ≠ f' ⟨x', hx'⟩) :
```

Comment:
Is this better started as `disjoint (set.range  f1) (set.range f2)`? Or does that just make things awkward?

Reply:
Hmm, it means you need `exact false.elim (im_disj ⟨⟨⟨x₁, h₁⟩, h⟩, ⟨⟨x₂, h₂⟩, rfl⟩⟩),` instead of `exact (im_disj h).elim`

Reply:
But then you could put this in `src/order/bounded_lattice` and then dot notation could be `range_disj.dite_injective hf hg`

Reply:
I suspect that this version could be harder to work with - but I think a reasonable compromise would just be to do both? `src/order/bounded_lattice` can have a lemma called `dite_injective` and have the statement you mention, and its proof can use this statement

Reply:
Something I'm wondering is whether the statement of the theorem should be in terms of sets for clarity:
```lean
lemma injective.dite_injective (s : set α) [decidable_pred s]
  (f : s → β) (f' : sᶜ → β)
  (hf : function.injective f) (hf' : function.injective f')
  (im_disj : ∀ {x x' : α} {hx : x ∈ s} {hx' : ¬ x' ∈ s}, f ⟨x, hx⟩ ≠ f' ⟨x', hx'⟩) :
  function.injective (λ x, if h : x ∈ s then f ⟨x, h⟩ else f' ⟨x, h⟩) := sorry
```
(This lemma is defeq to the original.)

Reply:
It's defeq, but I think the original pattern-matches better. A good test is to see which lemma can be rewritten by the other.

Reply:
> I suspect that this version could be harder to work with - but I think a reasonable compromise would just be to do both? `src/order/bounded_lattice` can have a lemma called `dite_injective` and have the statement you mention, and its proof can use this statement

I think I like this idea